### PR TITLE
Requestify regex pattern parsing

### PR DESCRIPTION
**Pull request by hamishknight**
Instead of doing the pattern parsing in both the C++ parser and ASTGen, factor out the parsing into a request that returns the pattern to emit, regex type, and version. This can then be lazily run during type-checking.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckRegex.cpp [hunk 1]
```diff
@@ -111,3 +133,62 @@ bool swift::decodeRegexCaptureTypes(ASTContext &ctx,
...
+                                         const RegexLiteralExpr *regex) const {
+#if SWIFT_BUILD_REGEX_PARSER_IN_COMPILER
+  auto &ctx = regex->getASTContext();
+  auto regexText = regex->getParsedRegexText();
+
+  // Let the Swift library parse the contents, returning an error, or null if
+  // successful.
+  size_t version = 0;
+  auto capturesBuf = ctx.AllocateUninitialized<uint8_t>(
+      getCaptureStructureSerializationAllocationSize(regexText.size()));
```
Comment (hamishknight) [inline lib/Sema/TypeCheckRegex.cpp:175]: Don't actually have to ASTContext allocate this anymore

**Comment by hamishknight** [comment]
@swift-ci please test

**Review by DougGregor — APPROVED** [review]
Looks good, thank you!

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test macOS

**Pull request merged**